Add support for --deny to exit with failure if there is an API diff

matthiasbeyer opened the pull request:
Closes #63 

This is merely a proposal.

I am not very sure about the actual flag name. `--check` seems to be ambiguous for me, maybe something along `--check-diff-empty` or `--is-empty` is more asppropriate.

Tell me what you think!

matthiasbeyer commented:
Not sure why CI fails, AFAICS, `std::process::ExitCode` implements `Termination`... ?

matthiasbeyer commented:
Hah, fun! `Result` does not implement `Termination` for `T: Termination`. I guess we have to work around that then, because I suspect returning `Result` from main is nice because of the error printing automatism.

Enselic commented on cargo-public-api/src/main.rs:
```diff
@@ -85,18 +85,30 @@ pub struct Args {
-        print_public_items_diff_between_two_commits(&args, commits)
+    let exit_success = if let Some(commits) = &args.diff_git_checkouts {
+        let there_is_diff = print_public_items_diff_between_two_commits(&args, commits)?;
+        args.check && !there_is_diff
```
CI fails because this returns `false`  even without `args.check` being true, which makes the process exit with failure. But if `args.check` is not true, diffing should always make the app exit with success.

I wonder if the code wouldn't become nicer if we instead of duplicating the code for `there_is_diff` handled this centrally? Since `args` is passed down, you can know deeper down if `args.check` is true or not.

I *think* I would be willing to call `std::process::exit(1)` deep down if that greatly simplifies the code. It is actually  more important that we write tests for what behaviour we want than exactly how we implement it. I sketched out what kind of tests I think we need in https://github.com/Enselic/cargo-public-api/issues/63#issuecomment-1167111370

matthiasbeyer replied:
> CI fails because this returns false even without args.check being true, which makes the process exit with failure. But if args.check is not true, diffing should always make the app exit with success.

Yeah, I added a patch simplifying the implementation, so that it is clearly visible what is happening.

Enselic reviewed (changes requested):
Thanks a lot for working on this! 🎉 

I made a high-level comment. I will wait with a detailed review until we have the big strokes in place.

Regarding if we should call the flag `--check`: I agree it does not feel like a super good fit. On the upside, it is the same name used in e.g. `rustfmt --check`. But I think it makes more semantic sense in there. `rustfmt --check` can be read as "check if formatting is OK".

In our case, `cargo public-api --diff-git-checkouts v0.1.0 v0.2.0 --check` would read as "check that the diff is OK between v0.1.0 and v0.2.0". That doesn't seem to make much sense.  Maybe call the flag `--deny-changes`? Besides, in the future, we might want to allow diffs to contain only added items but not removed items. In other words, allow users to have higher resolution on what diff that is allowed. Not sure what to call such an option though.

Another high level comment I have: It would be great if we could add tests to the code before we polish the implementation. Let me know if you want help with adding tests apart from the sketch I made in the comment I linked to.

matthiasbeyer commented:
>  Maybe call the flag --deny-changes? Besides, in the future, we might want to allow diffs to contain only added items but not removed items. In other words, allow users to have higher resolution on what diff that is allowed. Not sure what to call such an option though.

Yes, that's why I am particularly unhappy with `--check` :laughing: 

If we think about what a user might want:

* Check whether there is any diff
* Check whether there were things added
* Check whether there were things removed
* Check whether there were things changed

So I guess the flags should reflect that, whereas "check" would mean that they want a "successfull" exit if nothing is added/removed/changed - at least that's what I would want.

Maybe something like

* `--deny=all` (with the `=all` part being optional)
* `--deny=added`
* `--deny=removed`
* `--deny=changed`

matthiasbeyer commented:
Forgot to add:

>  It would be great if we could add tests to the code before we polish the implementation. Let me know if you want help with adding tests apart from the sketch I made in the comment I linked to.

Yes, definitively. Once we're settled I would like to add tests, and once we're settled I would also like to implement the other flags (so `--deny=all` but also `--deny=<all others>` :smile: ).

Enselic commented:
`--deny=all` or `--deny=added,removed` etc sounds good to me, at least as a first iteration!

Also feel free to only do `--deny=all` in this first PR. Not sure I think we should allow only `--deny` because that does not make semantic sense IMHO.

BTW your second set of CI failures was caused by an unrelated change to rustup released as 1.25.0 . I have merged a fix for that now via https://github.com/Enselic/cargo-public-api/pull/74 so if you rebase on that CI should become more happy

matthiasbeyer commented:
Rebased and added the changes we discussed. I would like to implement the "--deny=additions`/etc. variants in this PR as well (if you approve) because right now it does not make that much sense with only `all` as a possible value for `--deny` :laughing:

matthiasbeyer commented:
Does the above two patches for tests look good to you?

Enselic commented:
There is an [upcoming change](https://github.com/rust-lang/rust/pull/99287#pullrequestreview-1040717371) that might break this tool, so I need to take care of that with highest prio. I'll get back to this PR as soon as possible after that!

matthiasbeyer commented:
Sure, take your time! I am not in a hurry and won't go away! :laughing:

Enselic commented on cargo-public-api/src/main.rs:
```diff
@@ -85,18 +85,49 @@ pub struct Args {
-        print_public_items_diff_between_two_commits(&args, commits)
+    let exit_success = if let Some(commits) = &args.diff_git_checkouts {
+        let there_is_diff = print_public_items_diff_between_two_commits(&args, commits)?;
+        if args.deny.is_some() {
```
Here is how I approximately would like this code to be:

```rust
    let diff: Option<PublicItemsDiff> = if let Some(commits) = &args.diff_git_checkouts {
        print_public_items_diff_between_two_commits(&args, commits)?;
    } else if let Some(files) = &args.diff_rustdoc_json {
        print_public_items_diff_between_two_rustdoc_json_files(&args, files)?;
    } else {
        print_public_items_of_current_commit(&args)?;
        None
    };

    if let Some(diff) = diff {
        if args.deny.is_some() && diff.is_empty() {
            Err(Error::DiffDenied)?
        } else {
            Ok(())
        }
    } else {
        Ok(())
    }
```

which requires

```rust
#[derive(thiserror::Error, Debug)]
enum Error {
    #[error("THe API diff is not allowed as per --deny")]
    DiffDenied,
}
```

and also 

```toml
thiserror = "1.0.29"
```

in cargo-public-api/Cargo.toml

Note: Not polished, I just wanted to convey the idea.

matthiasbeyer replied:
Jup, done I think.

Enselic commented on cargo-public-api/src/main.rs:
```diff
@@ -85,18 +85,49 @@ pub struct Args {
+    All,
 }
 
-fn main_() -> Result<()> {
```
Rather than changing to `Result<bool>` which I find to be confusing, I think instead we should introduce an error. See other comments for more details.

matthiasbeyer replied:
Done.

Enselic commented on public-api/src/diff.rs:
```diff
@@ -106,6 +106,12 @@ impl PublicItemsDiff {
+
+    /// Check whether the diff is empty
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
```
Please add a call to this in existing tests in `tests/public-api-lib-tests.rs` and assert on that it returns the right value in all existing tests.

matthiasbeyer replied:
I think I did it the way you want :smile:

Enselic reviewed (changes requested):
Time to revisit this PR! 🎉  I added some more comments not to flesh out what kind of changes I would like to see. Let me know if you have any questions :) Looking forward to the next review round!

Probably a good idea to rebase on latest master BTW, because I refactored some of our tests recently.

To be clear: I am just making proposals, so don't interpret this as how it **must** look. Feel free to come up with alternatives, and we can discuss them :)

matthiasbeyer commented:
Here comes a rebase with the fixes you wanted (I hope I did not miss anything)!
Feel free to suggest another round of fixes if you want, I am looking forward!

matthiasbeyer commented:
FWIW, I'm going to mark the discussions I consider done as "completed", feel free to re-open them of course, if you think we can do better.

matthiasbeyer commented on cargo-public-api/src/main.rs:
```diff
@@ -85,18 +85,49 @@ pub struct Args {
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, clap::ArgEnum)]
+#[clap(rename_all = "lower")]
```
:+1: nice that I could teach you something :laughing:

matthiasbeyer commented on cargo-public-api/src/main.rs:
```diff
@@ -85,18 +85,49 @@ pub struct Args {
+    /// If specified with `--deny=all`, no diff is allowed to be there and cargo-public-api exits
+    /// with a nonzero exit code if there is.
+    #[clap(long, arg_enum)]
+    deny: Option<Vec<DenyMethod>>,
```
Sure. Done.

matthiasbeyer commented on cargo-public-api/src/main.rs:
```diff
@@ -230,6 +261,7 @@ fn main() -> Result<()> {
         },
-        result => result,
+        Ok(false) => std::process::exit(1),
+        Ok(true) => Ok(()),
```
I used a `if let Err(e)`, which is a bit more idomatic here I guess.

matthiasbeyer commented on cargo-public-api/tests/cargo-public-api-bin-tests.rs:
```diff
@@ -87,6 +87,28 @@ fn diff_public_items() {
+    cmd.arg("v0.3.0");
+    cmd.assert().success();
+}
+
```
I think I've done this change to your liking, feel free to nag me about this, of course!

Enselic commented on cargo-public-api/src/main.rs:
```diff
@@ -71,6 +72,15 @@ pub struct Args {
     #[clap(long, min_values = 2, max_values = 2)]
     diff_rustdoc_json: Option<Vec<String>>,
 
+    /// Deny diff
```
This is a bit too terse as this is the only thing told about the arg in "short help" mode (`cargo run -- -h`)

Enselic replied:
(fixed in https://github.com/Enselic/cargo-public-api/pull/72/commits/1ff21def29f5b07de3df91eba9f6691bd28469a9)

Enselic commented on cargo-public-api/src/main.rs:
```diff
@@ -92,15 +102,43 @@ pub struct Args {
+
+    if let Some(diff) = diff {
+        if args.deny.is_some() && diff.is_empty() {
+            anyhow::bail!(Error::DiffDenied)
```
`bail!` expands to `return`, but it is enough to just evaluate to what we want to return, since this is the last statement, so let's simplify

Enselic replied:
(fixed in https://github.com/Enselic/cargo-public-api/pull/72/commits/1ff21def29f5b07de3df91eba9f6691bd28469a9)

Enselic commented on cargo-public-api/src/main.rs:
```diff
@@ -231,13 +272,14 @@ or
 fn main() -> Result<()> {
-    match main_() {
-        Err(e) => match e.root_cause().downcast_ref::<std::io::Error>() {
+    if let Err(e) = main_() {
```
I started to think "maybe we don't need to change this at all", and seems like I was right, because if we leave this unchanged, the tool will simply say "`Error: The API diff is not allowed as per --deny`", which seems like a good thing to do. So let's just leave this code as-is.

Enselic replied:
(fixed in https://github.com/Enselic/cargo-public-api/pull/72/commits/1ff21def29f5b07de3df91eba9f6691bd28469a9)

Enselic reviewed (commented):
Thanks a lot for the update! That is a much better structuring of the code.

I had quite a few changes I would like still, so I figured it's easier for me to just push a commit (which I will do shortly after posting this comment) than making 20 small comments. I hope you don't mind. Do you think my changes look OK? Essentially: can you review my changes please? 😇 

Before merge I will squash commits together and remove my name from the code so that you will still get full credit for the change in git log and releases notes etc. Without you this wouldn't have happened, so you deserve the credit :)

Enselic commented on cargo-public-api/src/main.rs:
```diff
@@ -92,15 +102,43 @@ pub struct Args {
+
+// Actual main implementation
+//
+// Returns Ok(true) if everything went well
```
(fixed in https://github.com/Enselic/cargo-public-api/pull/72/commits/1ff21def29f5b07de3df91eba9f6691bd28469a9)

Enselic commented:
(When you do the review I recommend diffing against origin/main. The diff of my commit in isolation is probably confusing.)

matthiasbeyer commented on cargo-public-api/src/main.rs:
```diff
@@ -95,12 +103,34 @@ pub struct Args {
+    match (&args.deny, diff) {
+        // We were requested to deny diffs, so make sure there is no diff
+        (Some(_deny), Some(diff)) => {
+            if diff.is_empty() {
+                Ok(())
+            } else {
+                Err(anyhow!(error::Error::DiffDenied))
+            }
```
```suggestion
            diff.is_empty().then(|| Ok(())).unwrap_or_else(|| anyhow!(error::Error::DiffDenied))
```

What about (modulo some linebreaks maybe) :laughing:

Enselic replied:
Well, apart from that it is more complicated (`bool.then()` returns an `Option<T>`, but there is no need for that here I think) and does not compile, we could use that instead I guess 😅 

Why not use the current code though? Don't you think it is easier to see what happens? 

If you strongly prefer it to be something else I will for sure consider it, but I'd have to see something that compiles first :) (My deepest apologies if it does compile and I just misunderstood how to apply the change)

But I have to be honest: I strongly prefer simplicity over "smartness" (in lack of a better word).

matthiasbeyer replied:
> Why not use the current code though? Don't you think it is easier to see what happens?

I just like to explore all options in terms of "idomadicness" :smile: 

I'm okay with the current implementation of course, I consider this resolved.

matthiasbeyer commented on cargo-public-api/src/main.rs:
```diff
@@ -95,12 +103,34 @@ pub struct Args {
+        }
+
+        // We were requested to deny diffs, but we did not calculate a diff!
+        (Some(_), None) => Err(anyhow!("`--deny` can only be used when diffing")),
```
```suggestion
        (Some(_), None) => anyhow::bail!("`--deny` can only be used when diffing"),
```

Enselic replied:
Hmm not sure I'm a fan of that, because `anyhow::bail!` it expands to `return Err(anyhow!("`--deny` can only be used when diffing"))`, but there is no need to return early (i.e. "bail", because the evaluation of the expression is already what will be returned.

matthiasbeyer reviewed (commented):
I cannot block this PR because I am the author, but I have some minor things that could be improved.

Tell me if you want me to improve them or if you're okay with the current state and we can ignore these minor things.

And: Can I do the squashing please? You helped a great deal and I'd like to honor that in the commit! Just tell me if you only want one commit or whether several are okay!

Enselic commented:
Thanks a lot for code review! I responded in comments.

> And: Can I do the squashing please? You helped a great deal and I'd like to honor that in the commit! Just tell me if you only want one commit or whether several are okay!

Yes, of course! Feel free!

Normally I do regular git merges of all commits in a PR, but in this case I think there is a bit too much of fixup! and squash! commits to make sense to make them part of the persistent git log of the project

Doing the equivalent of `git rebase --autosquash` would probably be sufficient, but feel free to make any kind of cleanup you wish, as long as you use good judgement :)

matthiasbeyer commented:
Here we go.

I had to split your patch into several smaller ones to be able to squash them into the right commits properly. Next time please commit the individual changes `git add -p`. But no worries, I am some kind of git guru (some say so), so it wasn't a problem for me :laughing: 

I guess this is now ready for a final round of review and IMO also merging.
Make sure to have a look at the commit messages, whether I gave proper credit and the trailers are correct!

Besides that: :tada: 

Was a really pleasure to work with you on this! :heart:

Enselic reviewed (approved):
> Next time please commit the individual changes git add -p.

Yeah I'm actually quite good at that normally, but since in this case I planned to squash everything to one commit, I didn't bother :)

The new commit structure certainly looks good enough for merging as is, thanks!

I will let the PR sit for a day or two, because I've found that quite frequently you can think of some minor detail the hours after merging that makes you regret merging. So by waiting a day or two we can make some final adjustments if we think of any. Small things like typos and what not.

Plus, other maintainers gets a chance to review a bit too, if they want.

Yes, thanks for a good collaboration! 🎉 

Edit: Let's merge this now, I would like to push some code that depends on this PR being merged :P

Enselic commented:
For the record, this PR changes the `public-api` public API in the following way (as expected):


```
% cargo public-api --manifest-path public-api/Cargo.toml --diff-git-checkouts v0.12.3 main
Removed items from the public API
=================================
(none)

Changed items in the public API
===============================
(none)

Added items to the public API
=============================
+pub fn public_api::diff::PublicItemsDiff::is_empty(&self) -> bool
```

I mainly mention this because I like to diff the public api of public-api, because it's so meta. The secondary reason is that it is good to keep track of what PR makes what changes to the public API of public-api :)

Enselic commented:
Released as https://github.com/Enselic/cargo-public-api/releases/tag/v0.12.4 to crates.io 🥳

matthiasbeyer commented:
You released new features as a patch release?

Enselic commented:
In the Rust ecosystem, 0.X.Y is interpreted to mean that the major version is X and the minor version is Y. There is no number for patch versions unless you use 1.y.z in which case z is the patch number.

(Due to the rustdoc JSON format still changing in incompatible ways it is too early to enter 1.0.0 for this project IMHO).

See https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-cratesio for more info

> This compatibility convention is different from SemVer in the way it treats versions before 1.0.0. While SemVer says there is no compatibility before 1.0.0, Cargo considers 0.x.y to be compatible with 0.x.z, where y ≥ z and x > 0.

matthiasbeyer commented:
Well, you're talking about _compatibility_ here. I'm just talking about patch release vs. minor release. Cargo treats compatibility different than semver suggests, that does not mean that in `x.y.z` if `x` is `0` that `y` is "major". `x` is still major and `y` minor and `z` patch.

I was just wondering why you'd release a new feature with an increase in the "patch" position of the version string instead of a "minor", which would what semver defines. With your reasoning, it is understandable for me though.

Enselic commented:
If I would have released this as 0.13.0 I would have had to update https://github.com/Enselic/cargo-public-api#compatibility-matrix which I find to be too much of an effort TBH, plus it inflates version numbers unnecessarily (IMHO).

The most important thing for me is that nothing actually breaks. For example, if I remove an item from the public API and then release that without bumping the middle number, people will get build errors when they do cargo install cargo-public-api of a version that has not been updated accordingly. That would not be acceptable.

With this release, there should not be any breakage anywhere, which is the most important aspect.
